### PR TITLE
perf: BPF PID filter to break observer-effect feedback loop

### PR DESCRIPTION
**Pull request by adrian-mason**
## Summary
- Add `self_tgid` BSS variable to BPF probes, set by userspace before `load()` to wperf's own TGID — all 4 probe handlers early-return when any involved task matches wperf's TGID, breaking the observer-effect feedback loop (93% of events were self-observation)
- Zero-copy ringbuf write path: `write_event_raw()` passes raw BPF bytes directly to TLV stream, skipping from_bytes/to_bytes roundtrip
- Replace epoll-based `ringbuf.poll()` with `ringbuf.consume()` + explicit sleep — eliminates per-call epoll_wait kernel overhead
- Write path optimizations: 1 MiB BufWriter, single 45-byte write_all, 8192-event crash recovery interval
- Fix pending buffer final drain in perfarray path (Gemini-identified bug from PR #102)
- PID namespace correctness: `global_tgid()` reads NSpid from `/proc/self/status` for correct TGID in containers
- Spec amendment: §6.7 event rate threshold corrected from 100K to 10K (Oracle-approved — original included ~93% feedback-loop amplification)
- Add Directive 6 items to PR template (CI green, external reviews, owner approval)

## Authoritative Inputs
- Profiling data from Probe (perf record call-graph): 52% kernel scheduler → BPF PID filter; 4.35% epoll overhead → consume+sleep
- final-design.md §6.7 (overhead benchmark: <3% CPU gate, amended threshold)
- ADR-013 (dual-variant sched_switch + sched_wakeup probes)
- Oracle ruling: spec amendment 100K → 10K events/sec approved
- Oracle Directive 6: CI green + external reviews addressed + Adrian approval before merge

## Deviations
- None — self-filtering is standard profiler practice (OSDI'18 paper's WFG implicitly excludes the profiler). Consumption model changed from interrupt-driven to polling — acceptable for a recording tool that prioritizes completeness over latency.

## Dependency Checklist
- [x] No new dependencies

## Review Checklist
- [x] **Design conformance**: implementation matches the ADRs/specs listed in Authoritative Inputs
- [x] **Deviations declared**: any spec drift is explicitly listed above with a doc update reference
- [x] **Code correctness**: logic, error handling, edge cases
- [x] **Tests**: new/changed code has adequate test coverage
- [x] **CI green**: Coverage PASS, Supply Chain PASS, Check+Clippy+Test+Fmt — only failure is pre-existing bpftool/vmlinux.h gap (Oracle-exempted, tracked as Task #28)
- [x] **External reviews**: Gemini and all automated reviewer comments addressed (Directive 6) — BPF helper dedup, PID namespace, pending buffer drain all fixed
- [ ] **Owner approval**: @Adrian-Mason explicit merge authorization (Directive 6)

## Runtime Evidence

**W4 #23 PASS — 0.34% mean CPU (threshold <3.0%)**

Measured on commit `950044e` (all optimizations + Gemini fixes applied):

```json
{
  "gate": "W4 #23",
  "spec": "final-design.md §6.7",
  "threshold_cpu_percent": 3.0,
  "stressor": "matrix-64",
  "calibration_events_per_sec": 36795,
  "duration_seconds": 60,
  "num_runs": 3,
  "results": {
    "mean_cpu_percent": 0.34,
    "stddev_cpu_percent": 0.02,
    "total_events": 6607200,
    "total_drops": 0
  },
  "runs": [
    {"run":1, "cpu_percent":0.35, "events":2218710, "drops":0},
    {"run":2, "cpu_percent":0.32, "events":2197413, "drops":0},
    {"run":3, "cpu_percent":0.34, "events":2191077, "drops":0}
  ],
  "pass": true
}
```

Optimization journey (profile-guided):
| Config | Mean %CPU |
|--------|-----------|
| Baseline | 46.97% |
| + BPF PID filter | 4.15% |
| + write path opts | 3.89% |
| + consume+sleep | 0.36% |
| + Gemini fixes (final) | **0.34%** |

**Local BPF build verification (Oracle exemption condition #4):**
`cargo check --features bpf` — PASS on commit `0637218`

## Test Plan
- `cargo check --features bpf` passes locally
- `cargo test` — all unit tests pass
- Mutation testing: awaiting CI result (≥90% kill rate gate)
- Overhead measurement by Probe: 3 runs × 60s under stress-ng --matrix 64, mean 0.34% CPU, 0 drops
- Gemini feedback fixes verified: BPF helper dedup (commit 998e2b5), PID namespace + pending drain (commit 950044e)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/bpf/wperf.bpf.c [hunk 1]
```diff
@@ -127,6 +135,10 @@ int BPF_PROG(handle_sched_wakeup_btf,
 {
 	struct wperf_event *e;
 
+	if (self_tgid && (p->tgid == self_tgid ||
+			  (__u32)(bpf_get_current_pid_tgid() >> 32) == self_tgid))
```
[the remark this answers is not in the document]
Comment (adrian-mason) [reply src/bpf/wperf.bpf.c]: @gemini-code-assist Good catch — hoisted `bpf_get_current_pid_tgid()` before the filter check in both wakeup handlers (btf and raw_tp) so it's called once and reused for both the self-filter and event field population. Fixed in 998e2b5.

File: src/record.rs [hunk 2]
```diff
@@ -141,6 +141,13 @@ fn record_impl(args: &RecordArgs, stop_requested: &Arc<AtomicBool>) -> Result<()
         }
     }
 
+    open_skel
+        .maps
+        .bss_data
+        .as_mut()
+        .ok_or_else(|| RecordError::Bpf("BSS data not available".into()))?
+        .self_tgid = std::process::id();
```
[the remark this answers is not in the document]
Comment (adrian-mason) [reply src/record.rs]: @gemini-code-assist Valid concern for containerized use cases. For Phase 1, wperf requires root + BPF privileges and runs on the host (not in a PID namespace). The PID namespace scenario is worth tracking as a P2/P3 hardening item — if needed, we could read `/proc/self/status` NSpid field or use `bpf_get_current_pid_tgid()` from within a BPF program at init time. Filing as a follow-up.

[the remark this answers is not in the document]
Comment (adrian-mason) [reply src/record.rs]: @gemini-code-assist Update: we implemented the `global_tgid()` fix in commit 950044e — reads `/proc/self/status` NSpid field (first value = outermost TGID), with fallback to `std::process::id()`. This was promoted from P2 follow-up to a PR #106 fix after your review.

**Comment by adrian-mason** [comment]
@gemini-code-assist Re: poll timeout (500ms) and BufWriter seek flush — both are acceptable design tradeoffs for Phase 1. The 500ms timeout reduces CPU overhead while maintaining adequate shutdown responsiveness for a profiling tool. The seek-triggered flush in update_data_offset is infrequent (once at recording end) and correctness is more important than avoiding one extra flush. No changes needed.
